In [1]:
import numpy as np
from scipy.spatial import transform

In [2]:
def RotZ(theta_rad):
    return np.array([[np.cos(theta_rad), np.sin(theta_rad), 0],
                     [-np.sin(theta_rad), np.cos(theta_rad), 0],
                     [0, 0, 1]])
def RotY(theta_rad):
    return np.array([[np.cos(theta_rad), 0, -np.sin(theta_rad)],
                     [0, 1, 0],
                     [np.sin(theta_rad), 0, np.cos(theta_rad)]])
def RotX(theta_rad):
    return np.array([[1, 0, 0],
                     [0, np.cos(theta_rad), np.sin(theta_rad)],
                     [0, -np.sin(theta_rad), np.cos(theta_rad)]])

In [ ]:
"""Use Z-Y-X euler convention"""

### Question 1a

In [44]:
R_final  = RotX(30*np.pi/180)@RotY(-20*np.pi/180)@RotX(-30*np.pi/180)@RotY(20*np.pi/180)
R_final.round(3)

array([[ 0.984,  0.171, -0.043],
       [-0.17 ,  0.985,  0.034],
       [ 0.048, -0.026,  0.998]])

In [46]:
transform.Rotation.from_matrix(R_final).as_euler('ZYX')*180/np.pi

array([-9.77769559, -2.76359835, -1.4981269 ])

### Question 1b

The rotation matrix after the given transformation is given by:

In [20]:
R_current = (RotZ(-np.pi/2)@RotX(np.pi/2)@RotY(np.pi/2)).round(2)
R_current

array([[-1., -0., -0.],
       [ 0.,  0., -1.],
       [ 0., -1.,  0.]])

The rotation transformation matrix required to go back to original configuration:

In [47]:
R_req = R_current.T
R_req


array([[-1.,  0.,  0.],
       [-0.,  0., -1.],
       [-0., -1.,  0.]])

Confirming if we got the correct matrix. We should have $R_{\text{req}}R_{\text{current}} = \begin{bmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1 
\end{bmatrix}$

In [48]:
(R_req@R_current).round(2)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

To find what axis and by how much the vehicle needs to rotate to return to the original orientation, I will find the rotation vector representation of the $R_{\text{req}}$ matrix

In [49]:
rotvec = transform.Rotation.from_matrix(R_current.T).as_rotvec()
np.linalg.norm(rotvec), rotvec/np.linalg.norm(rotvec)

(3.141592653589793, array([ 0.        ,  0.70710678, -0.70710678]))

So we need to rotate about the [0, 0.707, -0.707] axis by $180 \degree$

The euler angles for final orientation are given by:

In [35]:
transform.Rotation.as_euler(transform.Rotation.from_matrix(R_current), 'ZYX')

array([ 3.14159265,  0.        , -1.57079633])

$$\phi = 180 \degree, \theta = 0 \degree, \psi = -90 \degree$$
Lets confirm this,

In [51]:
(RotX(-np.pi/2)@RotZ(np.pi)).round(2), R_current

(array([[-1.,  0.,  0.],
        [-0., -0., -1.],
        [-0., -1.,  0.]]),
 array([[-1., -0., -0.],
        [ 0.,  0., -1.],
        [ 0., -1.,  0.]]))

### Problem 2

We have $^B \omega = \begin{bmatrix}
0 \\
0 \\
r 
\end{bmatrix} $, Let $\underline{\theta_0}$ be the initial Euler Angles.
We know, $$^B \omega = E \underline{\dot \theta}$$, where $$E = \begin{bmatrix}
1 & 0 & -\sin \theta \\
0 & \cos \phi & \sin \phi \cos\theta \\
0 & -\sin\phi & \cos \phi \cos \theta 
\end{bmatrix}, \text{and } E^{-1} = \begin{bmatrix}
1 & \sin \phi \tan \theta & \cos\phi\tan\theta \\
0 & \cos\phi & -\sin\phi \\
0 & \sin\phi\sec\theta & \cos\phi\sec\theta 
\end{bmatrix}$$

Problem 2a

If this motion was superimposed over $\vec V  = V \hat i_E$,

Consider the equation for $\underline \theta$
$$ \underline{\dot\theta} = E^{-1} {^B\omega}$$
$$ \underline{\dot\theta} = \begin{bmatrix}
1 & \sin \phi \tan \theta & \cos\phi\tan\theta \\
0 & \cos\phi & -\sin\phi \\
0 & \sin\phi\sec\theta & \cos\phi\sec\theta 
\end{bmatrix} \begin{bmatrix}
0 \\
0 \\
r 
\end{bmatrix} = \begin{bmatrix} r \cos \phi \tan \theta \\ -r \sin \phi \\ r \cos\phi\sec\theta \end{bmatrix}$$